cdap-data-fabric/src/main/java/co/cask/cdap/spi/data/StructuredTable.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.spi.data;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.Range;
+
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ * Abstraction for a table that contains rows and columns.
+ * The schema of the table is fixed, and has to be specified in the
+ * {@link co.cask.cdap.spi.data.table.TableSpecification} during the table creation.
+ */
+public interface StructuredTable {
+  /**
+   * Write the collection of fields to the table.
+   * The fields contain both the primary key and the rest of the columns to write.
+   *
+   * @param fields the fields to write
+   * @throws InvalidFieldException if any of the fields are not part of the table schema, or the types do not match.
+   */
+  void write(Collection<Field<?>> fields) throws InvalidFieldException;
+
+  /**
+   * Read a single row from the table.
+   *
+   * @param keys the primary key of the row to read
+   * @param columns the columns to read if not all the columns are needed
```

Comment:
Does that mean an empty collection is for reading all columns? I think it is cleaner to have a separate method on this class for that, while passing in empty columns here would be an exception. This remove the confusion that an empty set is the same as a set with all column names.

Reply:
Added a separate method